### ISSUE TITLE
Diverting to same knot doesn't ++ the knot's count?

### ISSUE THREAD

**Issue by AdamNashGames** [inkle/ink#164]
I'm not sure if I'm misunderstanding the intended use or if this is a bug.  

It appears that when I divert to a knot that was the previous knot (knot ends with -> END and then another external ChoosePathString diverts to it again), the knot's {knot_name} variable does not increment.  

I'm currently storing each NPC as its own knot, and my Unity script is calling ChoosePathString to have the NPC talk.  I was hoping to use the knot's built-in variable to handle repeated interactions with the NPC. 

Note that when I divert to a different knot (i.e. speak to a different NPC), and then return to the first one, it DOES increment its count and behave as expected.  Also, I'm doing this all with one instance of Story, and global variables DO work as expected on repeated knots.

Perhaps I'm misunderstanding just how -> END works, especially with external ChoosePathString diverts.

**Comment by AdamNashGames** [comment]
Quick follow-up, in case this helps:

As a quick workaround, I've made each NPC divert to a TO_END knot that does nothing but divert to END, hoping that this would "register" the NPC's knot count.  

Turns out, it does, but only if there's some content there.  

Diverting to:
```
== TO_END
-> END
```

will not increment the previous knot's count, as my bug describes.  But doing anything at all, including:
```
== TO_END
~ TEST = TEST  // where TEST is a global VAR declared as 0 above
-> END
```

DOES work.  

In other words, diverting to an empty knot exhibits the same behavior, but diverting to a non-empty knot makes it work as expected (in the same way that talking to a different NPC works as described above).

**Comment by Fahrengeit** [comment]
Yeah, it was discussed some time before. Reason of this happening is, quoting:

> When diverting, the engine checks the difference in paths between where you've come from and where you're going, and only increments the read count for parts which change. The reason being for when you're jumping between different stitches within a knot, so that the read count for the knot doesn't increase each time.

So, maybe it will be solved, but workarounds are good enough for now.

**Comment by AdamNashGames** [comment]
Sorry that I missed that bit, and thanks for the help!  I'll go with the simple workaround for now.

**Comment by joethephish** [comment]
Hrm, yeah, it *should* have been solved last time, but it sounds like there might be something more subtle going on. I'm not managing to reproduce with this example:

```
== content_test ==
This is valid content
-> END

== empty_content_test ==
-> END

== count_stuff ==
content_test: {content_test}, empty_content_test: {empty_content_test}
-> END
```

C# code:

```
this.story.ChoosePathString ("count_stuff");
Console.Write ("Initial count: " + this.story.ContinueMaximally ());

this.story.ChoosePathString ("content_test");
Console.Write ("content_test: " + this.story.ContinueMaximally ());

this.story.ChoosePathString ("count_stuff");
Console.Write ("Count after content_test: " + this.story.ContinueMaximally ());

this.story.ChoosePathString ("empty_content_test");
Console.WriteLine ("empty_content_test: " + this.story.ContinueMaximally ());

this.story.ChoosePathString ("count_stuff");
Console.Write ("Count after empty_content_test: " + this.story.ContinueMaximally ());
```

(Output counts as you'd expect.)

Could you perhaps post a minimal repro?

**Comment by AdamNashGames** [comment]
I think your code works because it is actually using the workaround described above - you divert to another knot before reading the count again.  The "bug" (if it is one) I originally referred to was about diverting to the same thread multiple times in a row.  To repro, it's this simple:

Ink:
```
== TEST
Hello!  Count: {TEST}
-> END
```

C# (Unity): 
```
story.ChoosePathString("TEST");
Debug.Log(story.ContinueMaximally());

story.ChoosePathString("TEST");
Debug.Log(story.ContinueMaximally());
```

For example, imagine speaking to an NPC multiple times in a row in a Unity game that uses knots for NPCs and ChoosePathString to divert to that the knot when interacting with an in-game object.  If you want to advance the dialog to something new or even just add a Hello{NPC>1: again!}, you'd expect the knot count to increment each time ChoosePathString was called on it.  

Also, as mentioned above, a dummy divert at the end of every NPC knot is a very simple workaround.

**Comment by joethephish** [comment]
Ahha, I see! Okay, thanks I'll give that a try.

**Comment by joethephish** [comment]
Okay, so the reason this happens is that it caches the previous point in the story it was last at, so that doing something like:

```
== house ==
= room1
* go to room 2 -> room2

= room2
You're in room 2.
-> END
```

When you go from `room1` to `room2`, the address in the story changes from `house.room1` to `house.room2`. To prevent multiple visits to `house` (since you're already in the house), it *only updates the visit counts for the updated portion*. There are more complexities than that, but that's the gist.

So, of course, in your example, if the story state was still in your dialog knot when you immediately run it again, it thinks "you were already there", so the visit count doesn't need to be updated.

I experimented with a couple of different fixes, but found one that seems to be really simple, and works for your use case without hopefully affecting any other core functionality.

The new behaviour is that a call to `-> END` resets the cache of where you were before. That way, if the story state is in the middle of choice picking as in the example above, it will understand that you're effectively still "in the house". However, if `room1` had ended in `-> END`, then the location would be reset, and making a subsequent move into either `room2` or the `house` itself would trigger a new visit.

It would probably also be sensible to expose to C# the same functionality that `-> END` does internally, so you can effectively reset the run state (without affecting variable state etc), in case you want to cancel a story section before re-entering somewhere else (or indeed the same place again!).

It's a straightforward fix, so I'll push it to master right now if you want to grab the latest development build, otherwise it'll be in the next release.

**Comment by AdamNashGames** [comment]
Awesome!  Seems like a very intuitive fix.  I'd indeed expect -> END to sort of "exit" the "room", so that sounds like the perfect fix.  Of course, I'm completely unaware of other people's use-cases. :)

**Comment by joethephish** [comment]
Okay, it's in b82236085a4388dfc811eec0abd43275ef6e5c2a.
